yt_dlp/extractor/youtube.py

```diff
@@ -584,8 +587,200 @@ def _real_initialize(self):
         self._initialize_consent()
         self._check_login_required()
 
+    def _perform_login(self, username, password):
+        auth_type, sep, user = (username or '').partition('+')
+
+        if auth_type != 'oauth':
+            raise ExtractorError(
+                'Login using username and password is not supported. '
+                'Use  --username=oauth[+PROFILE] --password=""  to log in using oauth, '
+                f'or else {self._login_hint(method="cookies")}', expected=True)
+
+        self._initialize_oauth(user, password)
+
+    # OAuth 2.0 Device Authorization Grant flow, used by the YouTube TV client (youtube.com/tv).
+    #
+    # For more information regarding OAuth 2.0 and the Device Authorization Grant flow in general, see:
+    # - https://developers.google.com/identity/protocols/oauth2/limited-input-device
+    # - https://accounts.google.com/.well-known/openid-configuration
+    # - https://www.rfc-editor.org/rfc/rfc8628
+    # - https://www.rfc-editor.org/rfc/rfc6749
+    #
+    # Note: The official client appears to use a proxied version of the oauth2 endpoints on youtube.com/o/oauth2,
+    # which applies some modifications to the response (such as returning errors as 200 OK).
+    # Since the client works with the standard API, we will use that as it is well-documented.
+
+    _OAUTH_PROFILE = None
+    _OAUTH_ACCESS_TOKEN_CACHE = {}
+    _OAUTH_DISPLAY_ID = 'oauth'
+
+    # YouTube TV (TVHTML5) client. You can find these at youtube.com/tv.
+    _OAUTH_CLIENT_ID = '861556708454-d6dlm3lh05idd8npek18k6be8ba3oc68.apps.googleusercontent.com'
+    _OAUTH_CLIENT_SECRET = 'SboVhoG9s0rNafixCSGGKXAT'
+    _OAUTH_SCOPE = 'http://gdata.youtube.com https://www.googleapis.com/auth/youtube-paid-content'
+
+    # From https://accounts.google.com/.well-known/openid-configuration
+    # Technically, these should be fetched dynamically and not hard-coded.
+    # However, as these endpoints rarely change, we can risk saving an extra request for every invocation.
+    _OAUTH_DEVICE_AUTHORIZATION_ENDPOINT = 'https://oauth2.googleapis.com/device/code'  # device_authorization_endpoint
+    _OAUTH_TOKEN_ENDPOINT = 'https://oauth2.googleapis.com/token'                       # token_endpoint
+
+    @staticmethod
+    def _get_oauth_cache_key(profile):
+        return f'oauth_refresh_token_{profile}'
```

Comment:
maybe this would be nicer? `self._oauth_cache_key` vs `self._get_oauth_cache_key(profile)`
```suggestion
    @property
    def _oauth_cache_key(self):
        return f'oauth_refresh_token_{YoutubeBaseInfoExtractor._OAUTH_PROFILE}'
```